src/libs/TransactionUtils.js

```diff
@@ -268,34 +268,39 @@ function getAllReportTransactions(reportID) {
 }
 
 /**
- * Verifies that the provided waypoints are valid
+ * Filters the waypoints which are valid and returns those
  * @param {Object} waypoints
- * @returns {Boolean}
+ * @param {Boolean} reArrangeIndexes
+ * @returns {Object} validated waypoints
  */
-function validateWaypoints(waypoints) {
+function getValidWaypoints(waypoints, reArrangeIndexes = false) {
     const waypointValues = _.values(waypoints);
-
     // Ensure the number of waypoints is between 2 and 25
     if (waypointValues.length < 2 || waypointValues.length > 25) {
-        return false;
-    }
-
-    for (let i = 0; i < waypointValues.length; i++) {
-        const currentWaypoint = waypointValues[i];
-        const previousWaypoint = waypointValues[i - 1];
-
-        // Check if the waypoint has a valid address
-        if (!currentWaypoint || !currentWaypoint.address || typeof currentWaypoint.address !== 'string' || currentWaypoint.address.trim() === '') {
-            return false;
-        }
-
-        // Check for adjacent waypoints with the same address
-        if (previousWaypoint && currentWaypoint.address === previousWaypoint.address) {
-            return false;
-        }
+        return {};
     }
 
-    return true;
+    const validWaypoints = _.reduce(
+        waypointValues,
+        (acc, currentWaypoint, index) => {
+            const previousWaypoint = waypointValues[index - 1];
+            // Check if the waypoint has a valid address
+            if (!currentWaypoint || !currentWaypoint.address || typeof currentWaypoint.address !== 'string' || currentWaypoint.address.trim() === '') {
+                return acc;
+            }
+
+            // Check for adjacent waypoints with the same address
+            if (previousWaypoint && currentWaypoint.address === previousWaypoint.address) {
+                return acc;
+            }
+
+            const lastIndex = Math.max(..._.map(_.keys(acc), (key) => parseInt(key.replace('waypoint', ''), 10)));
+            return {...acc, [`waypoint${reArrangeIndexes && lastIndex > -1 ? lastIndex + 1 : index}`]: currentWaypoint};
+        },
+        {},
+    );
```

Comment:
I think we can refactor this code to be more readable.
How about changing to array first using `_.filter` and then back to object?
This won't require calculating key names in complex way.

Reply:
> Doing so will not preserve the keys. We want to not preserve the keys only when creating the distance request.

I also commented [here](https://github.com/Expensify/App/pull/26591#discussion_r1314237922). But did you find any issue without preserving keys on Request money page before creating request?

Reply:
What's the root cause? I thought `waypoints` is used to display routes, not `validatedWaypoints`

Reply:
https://github.com/Expensify/App/blob/388ddc8f157482d0b7a57d3dbcf016249b663e1d/src/components/DistanceRequest.js#L174

Reply:
ok, confirmed. I suspected `GetRoute` api would also require rearranged keys but no need

<img width="425" alt="response" src="https://github.com/Expensify/App/assets/108292595/c9873afa-3692-4439-b42b-7d8eadeaaacd">